// schemaValidator/test/sampleSchemas.js
'use strict';

const person = {
    name: { type: 'string', require: true, min: 5, max: 20 },
    firstName: { type: 'string', require: false, min: 5, max: 50 },
    age: { type: 'number', require: false, min: 3, max: 50 },
    info: { type: 'string', require: false, min: 10, max: 70 },
    birthday: { type: 'date', require: false },
    email: { type: 'email', require: false }
};

const schemaAddress = {
    street : { type: 'string', require: true, min: 5, max: 100 },
    zip: { type: 'string', require: true, min: 4, max: 8 },
    city : { type: 'string', require: true, min: 5, max: 100 }
};

const schemaComplete = {
    person: { type: 'schema', schema: 'person', require: true },
    address: { type: 'schema', schema: 'address', require: true }
};




const schemaBank = {
    bankName :  { type: 'string', require: true },
    bic : { type: 'string', require: true },
    konto: {
        type: 'schema',
        require: true,
        schema: {
            id: { type: 'number', require: true },
            iban: { type: 'string', require: true, min: 10 }
        }
    },
    inhaber: { type: 'string', require: true }
};

const arrayInt = {
    numbers: { type: 'array', require: true, min: 5, max: 10, valueMin: 2, valueMax: 15, arrayType: 'integer' }
};

const arrayString = {
    names: { type: 'array', require: true, min: 1, max: 5, valueMin: 3, valueMax: 10, arrayType: 'string' }
};

const arraySchema = {
    obj: { type: 'array', require: true, min: 1, max: 10, arrayType: 'schema', schema: 'person' }
};


const schemaEvents = {
    timeStamp : { type: 'date', require: true },
    text: { type: 'string', require: true }
};




module.exports = {
    person: person,
    schemaAddress: schemaAddress,
    schemaBank: schemaBank,
    arrayInt: arrayInt,
    arrayString: arrayString,
    arraySchema: arraySchema,
    schemaEvents: schemaEvents,
    schemaComplete: schemaComplete
};
